### PR TITLE
test(core): e2e for modal component

### PR DESCRIPTION
**Pull request by kulenkov**
# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [ ] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [ ] built and tested the changes locally.
 - [ ] added/updated tests.
 - [ ] added/updated API documentation.
 - [ ] added/updated demos.

### CODE REVIEW

File: cypress/full/modals_page_spec.ts [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+      });
+
+      it(`when user clicks on the button "Create template modalComponent" then modal is opened
+      "Close" button is present`, () => {
+        modals.clickByText(componentDemo, btnText);
+        modals.isModalVisible(modals.modalContainer, true);
+        modals.isModalBtnExist(modalBtnClose, 1);
+      });
+
+      it('user closes modal by clicking on "Close" button', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: when ...

File: cypress/full/modals_page_spec.ts [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+        modals.clickByText(eventsDemo, btnText);
+        modals.isModalVisible(modals.modalContainer, true);
+        modals.isModalDemoContainsText(eventsDemo, demoOnShowFired);
+        modals.isModalDemoContainsText(eventsDemo, demoOnShownFired, 1);
+      });
+
+      it(`when user closes modal by click on the cross then should be messages "onHide event has been fired"
+      and "onHidden event has been fired"`, () => {
+        // TODO: cy.get('card card-block card-header').eq(2).as('demo');
+
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: ?

File: cypress/support/modals.po.ts [hunk 3]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(`${baseSelector} ${modalSelector}`).eq(elementNumber).contains(btnTitle).click();
+  }
+
+  checkElementsQuantity(elemToCount: string, expectedQuantity: number) {
+    cy.get(elemToCount).should('have.length', expectedQuantity);
+  }
+
+  isModalDemoContainsText(baseSelector: string, expectedText: string, demoNumber = 0) {
+    cy.get(`${baseSelector} ${this.demoCardBlock}`).eq(demoNumber).invoke('text')
+      .should('contain', expectedText);
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: chai-jquery have func to.contain: https://docs.cypress.io/guides/references/assertions.html#Chai-jQuery
rewrite all your func like .should('to.contain', expectedText);

File: cypress/support/modals.po.ts [hunk 4]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  isModalBtnExist(btnTitle: string, elemNumber = 0) {
+    cy.get(this.modalContainer).find('button').eq(elemNumber).invoke('text')
+      .should('contain', btnTitle);
+  }
+
+  isDirectModalBtnExist(modalSelector: string, btnTitle: string, elemNumber = 0) {
+    cy.get(modalSelector).find('button').eq(elemNumber).invoke('text')
+      .should('contain', btnTitle);
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: change all such methods to "to.contains"

File: cypress/full/modals_page_spec.ts [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+        modals.clickByText(eventsDemo, btnText);
+        modals.clickOnModalBtn(btnX);
+        modals.isModalNotEnabled(modals.modalContainer, true);
+        cy.wait(500); // TODO: make without a wait
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: modals.clickByText(eventsDemo, btnText);
        modals.clickOnModalBtn(btnX);
        modals.isModalNotEnabled(modals.modalContainer, true);
        modals.isDemoContainsTxt(eventsDemo, demoOnHideFired);
        modals.isModalDemoContainsText(eventsDemo, demoOnHideFired, 2);
        modals.isModalDemoContainsText(eventsDemo, demoOnHiddenFired, 3);

File: cypress/full/modals_page_spec.ts [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+        modals.isModalDemoContainsText(eventsDemo, demoOnHideFired, 2);
+        modals.isModalDemoContainsText(eventsDemo, demoOnHiddenFired, 3);
+      });
+
+      it(`when user user closes modal by click outside the modal window then should be messages
+    "onHide event has been fired" and "onHidden event has been fired"`, () => {
+        modals.clickByText(eventsDemo, btnText);
+        modals.clickOutside(modals.modalContainer);
+        modals.isModalNotEnabled(modals.modalContainer, true);
+        cy.wait(500); // TODO: make without a wait
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: modals.clickByText(eventsDemo, btnText);
        modals.clickOutside(modals.modalContainer);
        modals.isModalNotEnabled(modals.modalContainer, true);
        modals.isBackdropDisabled();
        modals.isDemoContainsTxt(eventsDemo, demoOnHideFired);        modals.isModalDemoContainsText(eventsDemo, demoHideDismissed, 2);
        modals.isModalDemoContainsText(eventsDemo, demoHiddenDismissed, 3);

File: cypress/full/modals_page_spec.ts [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+
+      it(`when user closes modal by pressing ESC button then modal is closed and should be messages
+    "onHide event has been fired" and "onHidden event has been fired"`, () => {
+        // TODO: ESC does not work on a modal popup: 'cy.type() failed because it requires a valid typeable element'
+        // modals.clickByText(eventsDemo, btnText);
+        // cy.get(modals.modalContainer).type('{esc}');
+        // modals.isModalNotEnabled(modals.modalContainer, true);
+        // cy.wait(500); // TODO: make without a wait
+        // modals.isModalDemoContainsText(eventsDemo, demoOnEscDHide, 2);
+        // modals.isModalDemoContainsText(eventsDemo, demoOnEscHidden, 3);
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]:
``` 
modals.clickByText(eventsDemo, btnText);
        modals.pressEsc();
        modals.isModalNotEnabled(modals.modalContainer, true);
        modals.isBackdropDisabled();
        modals.isDemoContainsTxt(eventsDemo, demoOnEscDHide);
        modals.isModalDemoContainsText(eventsDemo, demoOnEscDHide, 2);
        modals.isModalDemoContainsText(eventsDemo, demoOnEscHidden, 3);```

File: cypress/full/modals_page_spec.ts [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+
+      it('example contains the buttons "Open modal" and "Disable animation"', () => {
+        modals.isButtonExist(animationDemo, btnText);
+        modals.isButtonExist(animationDemo, btnDisable, 1);
+      });
+
+      it('when user clicks on "Open modal" button then modal is opened. it appears with animations effects', () => {
+        modals.clickByText(animationDemo, btnText);
+        modals.isModalVisible(modals.modalContainer, true);
+        // can't check if the animation works
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: remove comment

File: cypress/full/modals_page_spec.ts [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+
+      it('example contains the buttons "Open modal" and "Disable Esc"', () => {
+        modals.isButtonExist(escapeDemo, btnText);
+        modals.isButtonExist(escapeDemo, btnDisable, 1);
+      });
+
+      it(`when user clicks on "Open modal" button then modal is opened. when user closes modal by click
+      ESC button then modal stays opened`, () => {
+        modals.clickByText(escapeDemo, btnText);
+        cy.get(modals.modalContainer).type('{esc}');
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: try to use         modals.pressEsc();

File: cypress/full/modals_page_spec.ts [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+      it(`when user clicks on "Disable Esc" button then title of the button is changed to "Enable animation"`,
+        () => {
+          modals.clickByText(escapeDemo, btnDisable);
+          modals.isButtonExist(escapeDemo, btnEnable, 1);
+        });
+
+      it(`after that click on "Open modal" button, modal popup is opened. when user press ESC button then modal
+      is closed`, () => {
+        modals.clickByText(escapeDemo, btnText);
+        cy.get(modals.modalContainer).type('{esc}');
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: modals.pressEsc();

File: cypress/full/modals_page_spec.ts [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+      it('when user clicks on "popup" button then a popup is shown', () => {
+        modals.clickByText(toolPopupDemo, btnText);
+        modals.clickOnModalBtn(btnPopover);
+        modals.isElementVisible('body', modals.modalPopup);
+      });
+
+      it(`when user hover on "tooltip" button then a popup is shown`,
+        () => {
+          modals.clickByText(toolPopupDemo, btnText);
+          cy.get(modals.modalContainer).contains('tooltip').focus();
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: move to PO or try to use any method from tooltip PO

File: cypress/full/modals_page_spec.ts [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+        modals.isButtonExist(backdropDemo, btnDisable, 1);
+        modals.isButtonExist(backdropDemo, btnDisableClick, 2);
+      });
+
+      it('when user clicks on "Open modal" button then modal is opened, it can be closed by clicking on a backdrop',
+        () => {
+          modals.clickByText(backdropDemo, btnText);
+          modals.isModalVisible(modals.modalContainer, true);
+          modals.isBackdropEnabled();
+          cy.get(`${'body'} ${modals.modalBackdrop}`).click({ force: true });
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: use methods from PO

File: cypress/full/modals_page_spec.ts [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+      });
+
+      describe('Optional sizes', () => {
+        beforeEach(() => modals.scrollToMenu('Optional sizes'));
+
+        const sizeDemo = modals.exampleDemosArr.directiveSizes;
+        const btnLargeModal = 'Large modal';
+        const btnSmallModal = 'Small modal';
+        const width300px = '300px';
+        const width800px = '800px';
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: in this describe better to set cy.viewport with appropriate width and heigth

File: cypress/full/modals_page_spec.ts [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+          modals.clickOnDirectModalBtn(nestedModalsDemo, modals.modalBody, btnOpen3rd, 1);
+          modals.isElemTextCorrect(nestedModalsDemo, modals.modalTitle, thirdModalTitle, 2);
+        });
+
+        it(`when user closes the third modal then the second is visible, and when user closes the second modal
+      then the first is visible`, () => {
+          modals.clickByText(nestedModalsDemo, btnText);
+          modals.clickOnDirectModalBtn(nestedModalsDemo, modals.modalBody, btnOpen2nd);
+          modals.clickOnDirectModalBtn(nestedModalsDemo, modals.modalBody, btnOpen3rd, 1);
+          cy.log('3rd modal is opened');
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: remove

File: cypress/full/modals_page_spec.ts [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+
+      it('example contains the button "Open a modal"', () => {
+        modals.isButtonExist(eventsModalsDemo, btnText);
+      });
+
+      it(`when user clicks on "Open a modal" button then modal is opened then should be two messages
+      "event onShow is fired" and "event onShown is fired"`, () => {
+        modals.clickByText(eventsModalsDemo, btnText);
+        modals.isDirectModalVisible(eventsModalsDemo, true);
+        cy.wait(500);
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: rework

File: cypress/full/modals_page_spec.ts [hunk 16]
```diff
@@ -0,0 +1,629 @@
...
+        modals.isModalDemoContainsText(eventsModalsDemo, eventOnShowFired);
+        modals.isModalDemoContainsText(eventsModalsDemo, eventOnShownFired, 1);
+      });
+
+      it(`when user closes modal by click on the cross then should be messages "event onHide is fired"
+      and "event onHidden is fired"`, () => {
+        modals.clickByText(eventsModalsDemo, btnText);
+        modals.clickOnDirectModalBtn(eventsModalsDemo, modals.crossSelector, btnX);
+        cy.wait(500); // TODO: make without a wait
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: rework

File: cypress/full/modals_page_spec.ts [hunk 17]
```diff
@@ -0,0 +1,629 @@
...
+        modals.isModalDemoContainsText(eventsModalsDemo, eventOnXHide, 1);
+        modals.isModalDemoContainsText(eventsModalsDemo, eventOnXHidden, 3);
+      });
+
+      it(`when user user closes modal by click outside the modal window then should be messages
+      "event onHidden is fired" and "onHidden event has been fired"`, () => {
+        modals.clickByText(eventsModalsDemo, btnText);
+        modals.clickOutside(eventsModalsDemo);
+        cy.wait(500); // TODO: make without a wait
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: rework

File: cypress/full/modals_page_spec.ts [hunk 18]
```diff
@@ -0,0 +1,629 @@
...
+      });
+
+      it(`when user user closes modal by pressing ESC button then modal is closed and should be messages
+      "event onHide is fired, dismissed by esc" and "event onHidden is fired, dismissed by esc"`, () => {
+        // TODO: ESC does not work on a modal popup: 'cy.type() failed because it requires a valid typeable element'
+        // modals.clickByText(eventsModalsDemo, btnText);
+        // cy.get(modals.modalContent).type('{esc}');
+        // cy.wait(500); // TODO: make without a wait
+        // modals.isModalDemoContainsText(eventsModalsDemo, eventOnOutsideHide, 2);
+        // modals.isModalDemoContainsText(eventsModalsDemo, eventOnOutsideHidden, 3);
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: rework

File: cypress/full/modals_page_spec.ts [hunk 19]
```diff
@@ -0,0 +1,629 @@
...
+        () => {
+          modals.clickByText(autoDemo, btnText);
+          modals.isDirectModalVisible(autoDemo, true);
+        });
+
+      it(`when user closes the modal then modal is removed from the DOM`,
+        () => {
+          modals.clickByText(autoDemo, btnText);
+          modals.clickOnDirectModalBtn(autoDemo, modals.crossSelector, btnX);
+          cy.get(`${autoDemo} ${modals.modalDialog}`).should('not.be.enabled');
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_page_spec.ts]: use methods from PO

File: cypress/integration/modals_page_spec.ts [hunk 20]
```diff
@@ -5,69 +5,40 @@ describe('Modals demo page test suite', () => {
...
+      modals.isModalVisible(modals.modalContainer, true);
+    });
 
-        cy.get(`${ childModals } ${ modals.backDirectiveMod }`).as('childModBack').click();
-        cy.get('@childModBack')
-          .should('not.to.be.visible');
-      });
+    it('when user press on ESC btn then the modal is closed', () => {
+      modals.clickByText(templateDemo, btnText);
+      cy.get(modals.modalContainer).type('{esc}');
```
Comment (ludmilanesvitiy) [inline cypress/integration/modals_page_spec.ts]: modals.pressEsc();

File: cypress/support/modals.po.ts [hunk 21]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
     directiveEvents: 'demo-modal-events',
     directiveAutoShow: 'demo-modal-auto-shown'
   };
+
+  // temporary placed here
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: ?

File: cypress/support/modals.po.ts [hunk 22]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  isElementVisible(baseSelector: string, elementToFind: string, elemNumber = 0) {
+    cy.get(`${baseSelector} ${elementToFind}`).eq(elemNumber).should('be.visible');
+  }
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: elementToFind => additionalSelector

Comment (ludmilanesvitiy) [reply cypress/support/modals.po.ts]: elemNumber => elementIndex

File: cypress/support/modals.po.ts [hunk 23]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+
+  isElemTextCorrect(baseSelector: string, itemSel: string, expectedText: string, rowNum = 0) {
+    cy.get(baseSelector).find(itemSel).eq(rowNum).invoke('text')
+      .should('contain', expectedText);
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: isItemTextContains(baseSelector: string, itemSelector: string, expectedText: string, rowNum = 0) {
    cy.get(baseSelector).find(itemSelector).eq(rowNum).invoke('text')
      .should('to.contains', expectedText);

File: cypress/support/modals.po.ts [hunk 24]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  isModalVisible(modalSelector: string, visible: boolean, elementNumber = 0) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: elementNumber =>contentNumber ?

File: cypress/support/modals.po.ts [hunk 25]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(`${'body'} ${modalSelector}`).find('.modal-content').eq(elementNumber)
+      .should(visible ? 'to.be.visible' : 'not.to.be.visible');
+  }
+
+  isDirectModalVisible(baseSelector: string, visible: boolean, elementNumber = 0) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: remove method, make above method - common for all cases

File: cypress/support/modals.po.ts [hunk 26]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(baseSelector).find('.modal-content').eq(elementNumber)
+      .should(visible ? 'to.be.visible' : 'not.to.be.visible');
+  }
+
+  isModalNotEnabled(modalSelector: string, disabled: boolean) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: enabled: boolean

Comment (ludmilanesvitiy) [reply cypress/support/modals.po.ts]: isModalNotEnabled => isModalEnabled

File: cypress/support/modals.po.ts [hunk 27]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(`${'body'} ${modalSelector}`).find('.modal-content')
+      .should(disabled ? 'not.to.be.enabled' : 'to.be.enabled');
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: rework according to comments above

File: cypress/support/modals.po.ts [hunk 28]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  isBackdropEnabled() {
+    cy.get('bs-modal-backdrop').should('to.have.class', 'show');
+  }
+
+  isBackdropDisabled() {
+    cy.get('bs-modal-backdrop').should('not.be.enabled');
+  }
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts:70]: combine this 2 methods into 1

File: cypress/support/modals.po.ts [hunk 29]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+
+  clickOnBackdrop() {
+    cy.get(`${'body'} ${this.modalBackdrop}`).click({ force: true });
+  }
+
+  isModalBtnExist(btnTitle: string, elemNumber = 0) {
+    cy.get(this.modalContainer).find('button').eq(elemNumber).invoke('text')
+      .should('contain', btnTitle);
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: remove method. reuse isButtonExist from baseComponent

File: cypress/support/modals.po.ts (hunk 4 above)
Comment (ludmilanesvitiy) [reply cypress/support/modals.po.ts]: reuse isButtonExist from baseComponent

File: cypress/support/modals.po.ts [hunk 30]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  isDirectModalBtnExist(modalSelector: string, btnTitle: string, elemNumber = 0) {
+    cy.get(modalSelector).find('button').eq(elemNumber).invoke('text')
+      .should('contain', btnTitle);
+  }
+
+  clickOnModalBtn(btnTitle: string) {
+    cy.get(this.serviceModalBtn).contains(btnTitle).click();
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: remove this method, reuse clickByText from baseCOmponent

File: cypress/support/modals.po.ts [hunk 31]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  clickOnDirectModalBtn(baseSelector: string, modalSelector: string, btnTitle: string, elementNumber = 0) {
+    cy.get(`${baseSelector} ${modalSelector}`).eq(elementNumber).contains(btnTitle).click();
+  }
+
+  checkElementsQuantity(elemToCount: string, expectedQuantity: number) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: elemToCount => elementsSelector

File: cypress/support/modals.po.ts [hunk 32]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(elemToCount).should('have.length', expectedQuantity);
+  }
+
+  isModalDemoContainsText(baseSelector: string, expectedText: string, demoNumber = 0) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: isPreviewExist from basecomponent

File: cypress/support/modals.po.ts [hunk 33]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+    cy.get(`${baseSelector} ${this.demoCardBlock}`).eq(demoNumber).invoke('text')
+      .should('contain', expectedText);
+  }
+
+  isModalTitleIs(modalTitle: string) {
+    cy.get('h4').should('contain', modalTitle);
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: selector here should be more specific

File: cypress/support/modals.po.ts [hunk 34]
```diff
@@ -26,4 +41,93 @@ export class ModalsPo extends BaseComponent {
...
+  }
+
+  isModalTooltipVisible() {
+    cy.get(this.modalContainer)
+      .should('to.have.descendants', this.modalTooltip)
+      .find('bs-tooltip-container')
+      .should('to.have.class', 'show');
+  }
+
+  isModalWindowWidth(modalSelector: string, expectedWidth: string, elementNumber = 0) {
```
Comment (ludmilanesvitiy) [inline cypress/support/modals.po.ts]: isModalWindowWidthEqual

**Review by ludmilanesvitiy — CHANGES_REQUESTED** [review]

File: cypress/support/modals.po.ts (hunk 21 above)
Comment (kulenkov) [reply cypress/support/modals.po.ts]: the method will be removed from here after the `e2e for typeahead` merged. the method is placed in `base.components`

File: cypress/full/modals_page_spec.ts (hunk 13 above)
Comment (kulenkov) [reply cypress/full/modals_page_spec.ts]: added to the `beforeEach()`

File: cypress/full/modals_directive_page.spec.ts [hunk 35]
```diff
@@ -0,0 +1,223 @@
+import { ModalsPo } from '../support/modals.po';
+
+describe('Modals demo page test suite', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_directive_page.spec.ts]: Modals demo page test suite: Directive examples

File: cypress/full/modals_directive_page.spec.ts [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+  const modals = new ModalsPo();
+
+  beforeEach(() => modals.navigateTo());
+
+  describe('Modals demo page test suite', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_directive_page.spec.ts]: remove

File: cypress/full/modals_directive_page.spec.ts [hunk 37]
```diff
@@ -0,0 +1,223 @@
...
+    beforeEach(() => modals.navigateTo());
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_directive_page.spec.ts]: remove

File: cypress/full/modals_directive_page.spec.ts [hunk 38]
```diff
@@ -0,0 +1,223 @@
...
+
+    describe('Directive examples', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_directive_page.spec.ts]: remove

File: cypress/full/modals_directive_page.spec.ts [hunk 39]
```diff
@@ -0,0 +1,223 @@
...
+      it(`when user user closes modal by pressing ESC button then modal is closed and should be messages
+      "event onHide is fired, dismissed by esc" and "event onHidden is fired, dismissed by esc"`, () => {
+        const modalText = 'Just another modal';
+
+        modals.clickByText(eventsModalsDemo, btnText);
+        modals.isDirectModalVisible(eventsModalsDemo, true);
+        modals.isItemTextContains(eventsModalsDemo, modals.modalBody, modalText);
+        modals.isBackdropExist(true);
+        modals.pressEsc();
+        modals.pressEscOnModal(eventsModalsDemo);
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_directive_page.spec.ts]: add modals.isBackdropExist(false);

File: cypress/full/modals_service_page_spec.ts [hunk 40]
```diff
@@ -0,0 +1,416 @@
+import { ModalsPo } from '../support/modals.po';
+import { TooltipPo } from '../support/tooltip.po';
+
+describe('Modals demo page test suite', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_service_page_spec.ts]: Modals demo page test suite: Service examples

File: cypress/full/modals_service_page_spec.ts [hunk 41]
```diff
@@ -0,0 +1,416 @@
...
+  const modals = new ModalsPo();
+
+  beforeEach(() => modals.navigateTo());
+
+  describe('Service examples', () => {
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_service_page_spec.ts]: remove

File: cypress/full/modals_service_page_spec.ts [hunk 42]
```diff
@@ -0,0 +1,416 @@
...
+        modals.isDemoContainsTxt(eventsDemo, demoOnHiddenFired);
+      });
+
+      it(`when user user closes modal by click outside the modal window then should be messages
+    "onHide event has been fired" and "onHidden event has been fired"`, () => {
+        modals.clickByText(eventsDemo, btnText);
+        modals.clickOutside(modals.modalContainer);
+        modals.isModalEnabled(modals.modalContainer, false);
+        modals.isBackdropExist(true);
+        modals.clickOnBackdrop();
```
Comment (ludmilanesvitiy) [inline cypress/full/modals_service_page_spec.ts]: add modals.isBackdropExist(false);

File: cypress/integration/modals_page_spec.ts [hunk 43]
```diff
@@ -5,69 +5,41 @@ describe('Modals demo page test suite', () => {
...
-      const buttonText = 'Open child modal';
-
-      it('directive child modal can be closed by backdrop click', () => {
-        modals.clickByText(childModals, buttonText);
-        cy.get(`${ childModals } ${ modals.modalContent }`)
-          .should('to.be.visible');
+    it('when user clicks on backdrop then the modal is closed', () => {
+      modals.clickByText(templateDemo, btnText);
+      modals.clickOnBackdrop();
+      modals.isModalVisible(modals.modalContainer, true);
```
Comment (ludmilanesvitiy) [inline cypress/integration/modals_page_spec.ts:37]: add modals.isModalEnabled(modals.modalContainer, false);
      modals.isBackdropExist(false);

File: cypress/integration/modals_page_spec.ts [hunk 44]
```diff
@@ -5,69 +5,41 @@ describe('Modals demo page test suite', () => {
...
-          .should('to.be.visible');
-
-        cy.get(`${ staticModal } ${ modals.modalHeader } ${ modals.btnCloseInHeader }`).click();
-        cy.get(`${ staticModal } ${ modals.backDirectiveMod }`)
-          .should('not.to.be.visible');
-      });
+    it('when user clicks on the cross button then the modal is closed', () => {
+      modals.clickByText(templateDemo, btnText);
+      modals.clickByText(modals.modalBtnX, btnX);
+      modals.isModalVisible(modals.modalContainer, true);
```
Comment (ludmilanesvitiy) [inline cypress/integration/modals_page_spec.ts:29]: add modals.isModalEnabled(modals.modalContainer, false);
      modals.isBackdropExist(false);

**Review by ludmilanesvitiy — CHANGES_REQUESTED** [review]

File: cypress/integration/modals_page_spec.ts (hunk 43 above)
Comment (kulenkov) [reply cypress/integration/modals_page_spec.ts:37]: got unexpected behavior with `modals.isBackdropExist(false);` - the backdrop does not disabled by click while running test in cypress, but it disabled on live site => i did not add this check

**Review by ludmilanesvitiy — APPROVED** [review]

**Pull request merged**